ledger/artifactmanager/blockhistory.go

```diff
@@ -0,0 +1,109 @@
+package artifactmanager
+
+import (
+	"context"
+	"errors"
+	"github.com/insolar/insolar/core"
+	"github.com/insolar/insolar/core/message"
+	"github.com/insolar/insolar/core/reply"
+	"github.com/insolar/insolar/ledger/record"
+)
+
+// HistoryIterator is used to iterate object history.
+//
+// During iteration object refs will be fetched from remote source.
+type HistoryIterator struct {
+	ctx        context.Context
+	messageBus core.MessageBus
+	object     core.RecordRef
+	chunkSize  int
+	fromPulse  *core.PulseNumber
+	fromChild  *core.RecordID
+	buff       []record.ObjectState
+	buffIndex  int
+	canFetch   bool
+}
+
+// History returns object's history references.
+func (d *ObjectDescriptor) History(pulse *core.PulseNumber) (core.RefIterator, error) {
+	return d.am.GetHistory(d.ctx, d.head, pulse)
+}
+
+// NewHistoryIterator creates new history iterator.
+func NewHistoryIterator(ctx context.Context, mb core.MessageBus, object core.RecordRef, fromPulse *core.PulseNumber, chunkSize int) (*HistoryIterator, error) {
+	iter := HistoryIterator{
+		ctx:        ctx,
+		messageBus: mb,
+		object:     object,
+		fromPulse:  fromPulse,
+		chunkSize:  chunkSize,
+		canFetch:   true,
+	}
+	err := iter.fetch()
+	if err != nil {
+		return nil, err
+	}
+	return &iter, nil
+}
+
+// HasNext checks if any elements left in iterator.
+func (i *HistoryIterator) HasNext() bool {
+	return i.hasInBuffer() || i.canFetch
+}
+
+// Next returns next element.
+func (i *HistoryIterator) Next() (*core.RecordRef, error) {
```

Comment:
We should probably return the data itself, not a reference. Object states do not have references. `.GetImage()` is either an object's prototype or a code reference. Neither serves our purpose.